atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractAtlasDbKeyValueServiceTest.java

```diff
@@ -660,21 +661,104 @@ public void testGetRangeOfTimestampsThrowsOnError() {
         Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
 
-        Value val = Value.create(data, TEST_TIMESTAMP + 1);
+        modifyReference(getRowsForCell(cell));
+
+        assertThat(getRowsForCell(cell), is(unmodifiedData));
+    }
+
+    @Test
+    public void testCannotModifyValuesAfterGet() {
+        Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
+
+        modifyReference(getForCell(cell));
+
+        assertThat(getForCell(cell), is(unmodifiedData));
+    }
+
+    @Test
+    public void testCannotModifyValuesAfterGetRange() {
+        Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
+
+        modifyReference(getOnlyItemInTableRange());
+
+        assertThat(getOnlyItemInTableRange(), is(unmodifiedData));
+    }
+
+    @Test
+    public void testCannotModifyValuesAfterGetRangeWithHistoryIfSupported() {
+        if (kvsSupportsGetRangeWithHistory()) {
+            testCannotModifyValuesAfterGetRangeWithHistory();
+        }
+    }
+
+    private boolean kvsSupportsGetRangeWithHistory() {
+        try {
+            keyValueService.getRangeWithHistory(TEST_TABLE, RangeRequest.all(), TEST_TIMESTAMP);
+            return true;
+        } catch (UnsupportedOperationException e) {
+            return false;
+        }
+    }
+
+    private void testCannotModifyValuesAfterGetRangeWithHistory() {
+        Cell cell = Cell.create(row0, column0);
+        byte[] unmodifiedData = writeToCell(cell, new byte[1]);
+
+        modifyReference(getOnlyItemInTableRangeWithHistory());
+
+        assertThat(getOnlyItemInTableRangeWithHistory(), is(unmodifiedData));
+    }
 
+    private void modifyReference(byte[] retrievedValue) {
+        retrievedValue[0] = (byte) 50;
+    }
+
+    private byte[] writeToCell(Cell cell, byte[] data) {
+        Value val = Value.create(data, TEST_TIMESTAMP + 1);
         keyValueService.putWithTimestamps(TEST_TABLE, ImmutableMultimap.of(cell, val));
+        return Arrays.copyOf(data, data.length);
+    }
 
-        data[0] = (byte) 50;
+    private byte[] getRowsForCell(Cell cell) {
+        return keyValueService.getRows(TEST_TABLE, ImmutableSet.of(cell.getRowName()), ColumnSelection.all(), TEST_TIMESTAMP + 3).get(cell).getContents();
+    }
 
-        assertThat(keyValueService.get(TEST_TABLE, ImmutableMap.of(cell, TEST_TIMESTAMP + 3)).get(cell).getContents(),
-                is(unmodifiedData));
+    private byte[] getForCell(Cell cell) {
+        return keyValueService.get(TEST_TABLE, ImmutableMap.of(cell, TEST_TIMESTAMP + 3)).get(cell).getContents();
+    }
 
-        keyValueService.delete(TEST_TABLE, ImmutableMultimap.of(cell, TEST_TIMESTAMP + 1));
+    private byte[] getOnlyItemInTableRange() {
+        ClosableIterator<RowResult<Value>> rangeIterator = keyValueService.getRange(TEST_TABLE, RangeRequest.all(), TEST_TIMESTAMP + 3);
```

Comment:
need to close these iterators to avoid resource leaks, would suggest try-with-resources and wrapping lines.